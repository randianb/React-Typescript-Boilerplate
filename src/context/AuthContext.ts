import React from 'react';

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
