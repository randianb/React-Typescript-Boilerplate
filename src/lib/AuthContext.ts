import React from 'react';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
