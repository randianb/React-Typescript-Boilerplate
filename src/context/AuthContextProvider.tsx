import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';
import { withRouter, RouteComponentProps } from 'react-router';
import auth0 from 'auth0-js';

export interface AuthContextProviderConfig {
  responseType?: string;
  scope?: string;
}

export interface AuthContextProviderProps extends RouteComponentProps {
  config?: AuthContextProviderConfig;
  loginCallback?: () => void;
  logoutCallback?: () => void;
  renewSessionCallback?: () => void;
}

class AuthContextProviderComponent extends React.Component<
  AuthContextProviderProps
> {
  private auth0: auth0.WebAuth;
  private tokenRenewalTimeout: NodeJS.Timeout | null = null;

  constructor(props: AuthContextProviderProps) {
    super(props);

    if (!process.env.REACT_APP_AUTH_DOMAIN) {
      throw new Error('missing REACT_APP_AUTH_DOMAIN config');
    }

    if (!process.env.REACT_APP_AUTH_CLIENT_ID) {
      throw new Error('missing REACT_APP_AUTH_CLIENT_ID config');
    }

    if (!process.env.REACT_APP_AUTH_CALLBACK) {
      throw new Error('missing REACT_APP_AUTH_CALLBACK config');
    }

    this.auth0 = new auth0.WebAuth({
      domain: process.env.REACT_APP_AUTH_DOMAIN,
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID,
      redirectUri: process.env.REACT_APP_AUTH_CALLBACK,
      responseType:
        (props.config && props.config.responseType) || 'token id_token',
      scope: (props.config && props.config.scope) || 'openid'
    });
  }

  public state: AuthContextInterface = {
    token: null,
    idToken: null,
    expiresAt: 0,
    checkingSession: false,
    isAuthenticated: false,
    handleAuthentication: this.handleAuthentication.bind(this),
    login: this.login.bind(this),
    logout: this.logout.bind(this)
  };

  public componentDidMount() {
    if (localStorage.getItem('isLoggedIn') === 'true') {
      this.renewSession();
    }
  }

  public login() {
    this.auth0.authorize();
  }

  public handleAuthentication() {
    const { loginCallback } = this.props;
    this.auth0.parseHash(async (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        await this.setSession(authResult);
        if (loginCallback) {
          loginCallback();
        }
      }
      this.props.history.push('/');
    });
  }

  public getToken() {
    return this.state.token;
  }

  public getIdToken() {
    return this.state.idToken;
  }

  public async setSession(authResult: auth0.Auth0DecodedHash): Promise<void> {
    localStorage.setItem('isLoggedIn', 'true');

    const expiresAt = authResult.expiresIn
      ? authResult.expiresIn * 1000 + new Date().getTime()
      : 0;

    await new Promise(resolve => {
      this.setState(
        {
          token: authResult.accessToken || null,
          idToken: authResult.idToken || null,
          expiresAt,
          isAuthenticated: true,
          checkingSession: false
        },
        () => {
          this.scheduleRenewal();
          resolve();
        }
      );
    });
  }

  public renewSession() {
    const { renewSessionCallback } = this.props;

    this.setState({
      checkingSession: true
    });

    this.auth0.checkSession({}, async (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        await this.setSession(authResult);
        if (renewSessionCallback) {
          renewSessionCallback();
        }
      } else if (err) {
        this.logout();
      }
    });
  }

  public logout() {
    const { logoutCallback } = this.props;

    if (this.tokenRenewalTimeout) {
      clearTimeout(this.tokenRenewalTimeout);
    }

    this.setState(
      {
        token: null,
        idToken: null,
        expiresAt: 0,
        isAuthenticated: false,
        checkingSession: false
      },
      () => {
        if (logoutCallback) {
          logoutCallback();
        }
      }
    );

    localStorage.removeItem('isLoggedIn');

    this.auth0.logout({
      returnTo: window.location.origin
    });
  }

  public isAuthenticated() {
    const expiresAt = this.state.expiresAt || 0;
    return new Date().getTime() < expiresAt;
  }

  public scheduleRenewal() {
    const expiresAt = this.state.expiresAt || 0;
    const timeout = expiresAt - Date.now();

    if (timeout > 0) {
      this.tokenRenewalTimeout = setTimeout(() => {
        this.renewSession();
      }, timeout);
    }
  }

  public render() {
    return (
      <AuthContext.Provider value={this.state}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AppContextProvider = withRouter(AuthContextProviderComponent);
