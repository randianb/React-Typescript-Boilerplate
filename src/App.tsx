import React from 'react';
import {
  Route,
  RouteComponentProps,
  withRouter,
  Switch,
  Redirect
} from 'react-router-dom';
import Callback from './pages/Callback/Callback';
import Admin from './pages/Admin/Admin';
import Login from './pages/Login/Login';
import Logout from './pages/Logout/Logout';
import SecuredRoute from './components/SecuredRoute/SecuredRoute';
import Nav from './components/Nav/Nav';
import './app.scss';
import Home from './pages/Home/Home';
import { withAuth, AuthProps } from './lib/with-auth';

interface AppProps extends AuthProps, RouteComponentProps<any> {}

class App extends React.Component<AppProps> {
  public componentDidMount() {
    const { auth } = this.props;

    if (this.props.location.pathname === '/callback') {
      auth.checkingSession = false;
      this.forceUpdate();
      return;
    }

    // check to see if user is already authed
    auth
      .silentAuth()
      .then(() => {
        auth.checkingSession = false;
        this.forceUpdate();
      })
      .catch(err => {
        auth.checkingSession = false;
        this.forceUpdate();
        if (err.error !== 'login_required') {
          console.log(err.error);
        }
      });
  }

  public render() {
    return (
      <div>
        <Nav />
        <Switch>
          <Route exact={true} path="/login" component={Login} />
          <Route exact={true} path="/logout" component={Logout} />
          <Route exact={true} path="/callback" component={Callback} />
          <SecuredRoute exact={true} path="/admin" component={Admin} />
          <Route exact={true} path="/" component={Home} />
          <Redirect to="/" />
        </Switch>
      </div>
    );
  }
}

export default withAuth(withRouter(App));
