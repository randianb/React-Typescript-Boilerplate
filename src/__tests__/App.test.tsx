import React from 'react';
import App from '../App';
import { render, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { authContextDefaults, AuthContext } from '../context/AuthContext';

afterEach(cleanup);

describe('App', () => {
  const auth = authContextDefaults;
  const TestComponent = (props: any) => {
    return (
      <AuthContext.Provider value={props.auth}>
        <MemoryRouter
          initialEntries={props.paths}
          initialIndex={props.index || 0}
        >
          <App />
        </MemoryRouter>
      </AuthContext.Provider>
    );
  };

  it('Renders App home', () => {
    const { container } = render(
      <TestComponent auth={auth} paths={['/']} index={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders Admin', () => {
    auth.isAuthenticated = true;
    const { container } = render(
      <TestComponent auth={auth} paths={['/admin']} index={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders login', () => {
    const { container } = render(
      <TestComponent auth={auth} paths={['/login']} index={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders logout', () => {
    const { container } = render(
      <TestComponent auth={auth} paths={['/logout']} index={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders callback', () => {
    const { container } = render(
      <TestComponent auth={auth} paths={['/callback']} index={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
