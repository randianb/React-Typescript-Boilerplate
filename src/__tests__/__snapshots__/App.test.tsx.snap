// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders Admin 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      Home
    </a>
    <a
      href="/admin"
    >
      Admin
    </a>
    <a
      href="/logout"
    >
      Logout
    </a>
  </div>
  <main
    class="admin"
  >
    <h1>
      Admin
    </h1>
  </main>
</div>
`;

exports[`App Renders App home 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      Home
    </a>
    <a
      href="/admin"
    >
      Admin
    </a>
    <a
      href="/login"
    >
      Login
    </a>
  </div>
  <main
    class="home"
  >
    <h1>
      Welcome
    </h1>
  </main>
</div>
`;

exports[`App Renders callback 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      Home
    </a>
    <a
      href="/admin"
    >
      Admin
    </a>
    <a
      href="/logout"
    >
      Logout
    </a>
  </div>
  <div
    class="callbackPage"
  >
    <div
      class="Spinner"
    >
      <div
        class="Spinner-spinner"
      >
        <svg
          viewBox="0 0 64 64"
        >
          <circle
            r="26"
            transform="translate(32,32)"
          />
        </svg>
      </div>
      <div
        class="Spinner-content"
      >
        Loading Your Profile
      </div>
    </div>
  </div>
</div>
`;

exports[`App Renders login 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      Home
    </a>
    <a
      href="/admin"
    >
      Admin
    </a>
    <a
      href="/logout"
    >
      Logout
    </a>
  </div>
  <div
    class="Spinner loginPage"
  >
    <div
      class="Spinner-spinner"
    >
      <svg
        viewBox="0 0 64 64"
      >
        <circle
          r="26"
          transform="translate(32,32)"
        />
      </svg>
    </div>
    <div
      class="Spinner-content"
    >
      Logging in
    </div>
  </div>
</div>
`;

exports[`App Renders logout 1`] = `
<div>
  <div
    class="navbar"
  >
    <a
      href="/"
    >
      Home
    </a>
    <a
      href="/admin"
    >
      Admin
    </a>
    <a
      href="/logout"
    >
      Logout
    </a>
  </div>
  <main
    class="home"
  >
    <h1>
      Welcome
    </h1>
  </main>
</div>
`;
